File: js/05-emparejando-botas.js
function organizeShoes(shoes) {
  const pairOfShoes = [];
  const shoeCounts = new Map();

  for (const { type, size } of shoes) {
    const oppositeType = type === 'I' ? 'R' : 'I';
    const oppositeKey = `${oppositeType}_${size}`;
    const currentKey = `${type}_${size}`;

    if (shoeCounts.get(oppositeKey) > 0) {
      shoeCounts.set(oppositeKey, shoeCounts.get(oppositeKey) - 1);
      pairOfShoes.push(size);
    } else {
      shoeCounts.set(currentKey, (shoeCounts.get(currentKey) || 0) + 1);
    }
  }

  return pairOfShoes;
}
const shoes = [
  { type: 'I', size: 38 },
  { type: 'R', size: 38 },
  { type: 'R', size: 38 }
];
console.log(organizeShoes(shoes));

File: python/05-emparejando-botas.py
def organize_shoes(shoes: list[dict]) -> list[int]:
    pair_of_shoes = []
    if not shoes:
        return pair_of_shoes

    map_of_shoes = {}

    for shoe in shoes:
        opposite_shoe = 'R' if shoe['type'] == 'I' else 'I'
        opposite_pair = f"{opposite_shoe}_{shoe['size']}"

        if map_of_shoes.get(opposite_pair, 0) > 0:
            map_of_shoes[opposite_pair] -= 1
            pair_of_shoes.append(shoe['size'])
        else:
            current_shoe = f"{shoe['type']}_{shoe['size']}"
            map_of_shoes[current_shoe] = map_of_shoes.get(current_shoe, 0) + 1

    return pair_of_shoes